platform/cms/accelerated_publishing/rfc/broken-link-checking.md

```diff
@@ -0,0 +1,126 @@
+2. The speed gains of Next Build / Accelerated Publishing will come from "incremental publishing". Put simply, Content Build operates by publishing every single CMS page every single publishing run. Most of this publishing is not useful, because nothing has changed in the content. With Next Build, we will transition to a paradigm where we only build pages that have changes. This should drastically reduce the amount of pages that a single publishing run requires. However this also means these pages will be built in small batches. This also presents problems for points 2. and 3. above.
+3. Finally, link checking can be a time-consuming process. With our stated goal of increasing publishing speed and being able to expand the amount of content the CMS can maintain, we want to avoid adding time to publishing. Rather than checking links at build time and having it add to that build time, it would be good to test for broken links independently.
+
+### Problems to solve
+
+In summary, the problems that the current system has are:
+
+- Content Build will no longer have complete knowledge of all pages, and so any pages that have migrated to Next Build will be flagged as broken links by Content Build
+- We will still have to test for broken links for the site as a whole.
+- Running full broken links checks at build time is too slow for an accelerated publishing scenario.
+
+## Design
+
+_Explain the proposed changes in enough detail so that a team member with working knowledge
+of the project could implement the change themselves. Bonus points if this is easily mappable
+to user stories or specs._
+
+### Link Checking Features
+
+Our system must do the following
+
+- It must operate on the entirety of va.gov content, regardless of what system the page was built by.
+- It must not create an impediment to the build process.
+- It must report the broken links into a central location that can respond to the issues.
+
+### Crawler Proposal
+
+A straight-forward way would be to create a site crawler that checks va.gov pages, follows any links found, and reports any that return a non-200/300 response. This would operate as a CI process on VA.gov's published content.
+
+All content published by Content Build and Next Build is listed in sitemap.xml files. These URLs can be read by a site crawler and the contents of the URL scanned.
+
+Advantages:
+
+- This would be relatively simple to create.
+- May be able to be largely addressed by of-the-shelf software. or entirely. [The Node.js ecosystem has several broken link checking packages](https://www.npmjs.com/search?q=broken%20link), for example.
+- Both internal and external links would be collected together into one report.
+
+Challenges:
+
+- Content would no longer be checked as it is built, which is an expectation VA.gov has due to the current architecture.
+- Determining how frequently to scan will require some experimentation.
```

Comment:
Scanning once per hour if it can handled it ought to be sufficient to cover existing build.  Though it may function better as the way we currently do content releases.  When one full scan completes, it kicks off the next.
Would actually want it to run a bit on the slow side so that we don't have race conditions where someone creates an event  (next-build) with a link to a content build based page.  The event would land first with a link to a page that has not been built yet by content-build.

Reply:
I think trying to account for race conditions with respect to content publishing/availability between next-build and content-build is a losing game, frankly. The only way to ensure that the two are in sync would be to yoke next-build to content-build; to only allow next-build to release when content-build releases. This effectively would mean we see no speed improvements of any kind until content-build is 100% retired.

I *do* think that there will need to be an education process for CMS User Support in diagnosing broken link issues, since there will be some 'intermittent' broken link issues resulting from them not being yoked together. I.e., a full site scan might report a broken link from a next-build page to a content-build page because content-build is lagging behind. I think this would need to become part of the assessment/triage process for broken link response that currently is happening.

See also the comment below about notifications, a potential broken link dashboard, etc.